### PR TITLE
test: add functional test for agent introspection endpoint

### PR DESCRIPTION
**Pull request by haikuoliu**
### Summary

Add functional test for agent introspection endpoint.

### Implementation details

The `AgentIntrospectionValidator` basically mimics `TaskMetadataValidator`:
1. Start a container to query the agent introspection endpoint, the container is started with `host` mode, so it's similar to the case when you ssh into the container instance and query the introspection endpoint.
2. In the container, the tasks metadata and task metadata path are queried, they are unmarshalled to v1 response struct, since the the value of the fields are dynamic, for the most of them we only check if they are empty.
3. The Agent metadata is already checked when we start Agent in the functional test, so we don't check it in the container.

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/functional_tests/testdata/taskdefinitions/agent-introspection-validator/task-definition.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+  "family": "ecsftest-agent-introspection-validator",
+  "networkMode": "host",
+  "containerDefinitions": [{
+    "image": "amazon/amazon-ecs-agent-introspection-validator:make",
+    "name": "agent-introspection-validator",
+    "memory": 50,
+    "healthCheck": {
```
Comment (adnxn) [inline agent/functional_tests/testdata/taskdefinitions/agent-introspection-validator/task-definition.json]: does this test need to have a healthcheck?

File: agent/functional_tests/testdata/taskdefinitions/agent-introspection-validator/task-definition.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        "command": ["CMD-SHELL", "echo hello"],
+        "interval": 5,
+        "timeout": 5,
+        "retries": 2,
+        "startPeriod": 1
+    },
+    "logConfiguration": {
+        "logDriver": "awslogs",
```
Comment (adnxn) [inline agent/functional_tests/testdata/taskdefinitions/agent-introspection-validator/task-definition.json]: why are we configuring this task to use awslogs?

File: misc/agent-introspection-validator/agent-introspection-validator.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+}
+
+// TasksResponse is the schema for the tasks response JSON object
+type TasksResponse struct {
+	Tasks []*TaskResponse
+}
+
+// ContainerResponse is the schema for the container response JSON object
+type ContainerResponse struct {
+	DockerID   string
```
Comment (adnxn) [inline misc/agent-introspection-validator/agent-introspection-validator.go]: just making a note since i asked you offline, this should be `DockerId` to match the original API.

File: agent/functional_tests/testdata/taskdefinitions/agent-introspection-validator/task-definition.json (hunk 1 above)
Comment (haikuoliu) [reply agent/functional_tests/testdata/taskdefinitions/agent-introspection-validator/task-definition.json]: No, this is unnecessary, I will remove it.

File: agent/functional_tests/testdata/taskdefinitions/agent-introspection-validator/task-definition.json (hunk 2 above)
Comment (haikuoliu) [reply agent/functional_tests/testdata/taskdefinitions/agent-introspection-validator/task-definition.json]: Because the logs in `misc/agent-introspection-validator/agent-introspection-validator.go` needs be to be uploaded to Cloud Watch.

File: misc/agent-introspection-validator/agent-introspection-validator.go (hunk 3 above)
Comment (haikuoliu) [reply misc/agent-introspection-validator/agent-introspection-validator.go]: Thanks! I was planning to update it when the fix PR is rebased, will add a TODO in the PR description.

File: misc/agent-introspection-validator/agent-introspection-validator.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+	IPv6Addresses []string `json:"IPv6Addresses,omitempty"`
+}
+
+func getTasksMetadata(client *http.Client, path string) (*TasksResponse, error) {
+	body, err := metadataResponse(client, path, tasksMetadataRespType)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Printf("Received tasks metadata: %s \n", string(body))
```
Comment (yunhee-l) [inline misc/agent-introspection-validator/agent-introspection-validator.go]: should these fmt.Printfs be logging statements?

File: misc/agent-introspection-validator/agent-introspection-validator.go [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+func getTasksMetadata(client *http.Client, path string) (*TasksResponse, error) {
+	body, err := metadataResponse(client, path, tasksMetadataRespType)
+	if err != nil {
+		return nil, err
+	}
+
+	seelog.Infof("Received tasks metadata: %s \n", string(body))
+
+	var tasksMetadata TasksResponse
+	err = json.Unmarshal(body, &tasksMetadata)
```
Comment (richardpen) [inline misc/agent-introspection-validator/agent-introspection-validator.go]: Unmarshal could also work with case insensitive, this won't capture the bug, example: https://play.golang.org/p/iH8nsK-UJ_L

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 6]
```diff
@@ -609,6 +609,48 @@ func TestMetadataServiceValidator(t *testing.T) {
 	assert.Equal(t, 42, exitCode, fmt.Sprintf("Expected exit code of 42; got %d", exitCode))
 }
 
+// TestAgentIntrospectionValidator tests that the agent introspection endpoint can
+// be accessed from within the container.
+func TestAgentIntrospectionValidator(t *testing.T) {
+	// The Docker version was 17.03.1-ce when we added changes to agent
+	// introspection endpoint feature for the last time.
+	RequireDockerVersion(t, ">=17.03.1-ce")
```
Comment (richardpen) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Why was this required? The introspection API shouldn't have dependency on the docker version.

File: agent/functional_tests/tests/functionaltests_unix_test.go [hunk 7]
```diff
@@ -609,6 +609,48 @@ func TestMetadataServiceValidator(t *testing.T) {
...
+			return
+		}
+		task.WaitStopped(waitTaskStateChangeDuration)
+	}()
+
+	err = task.WaitStopped(waitTaskStateChangeDuration)
+	require.NoError(t, err, "Error waiting for task to transition to STOPPED")
+	exitCode, _ := task.ContainerExitcode("agent-introspection-validator")
+
+	assert.Equal(t, 42, exitCode, fmt.Sprintf("Expected exit code of 42; got %d", exitCode))
```
Comment (richardpen) [inline agent/functional_tests/tests/functionaltests_unix_test.go]: Minor: you can just use `assert.Equalf` to get rid of the `fmt.Sprintf`.

File: misc/agent-introspection-validator/agent-introspection-validator.go [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+
+	var actualContainerName string
+	json.Unmarshal(containerMetadataMap["Name"] ,&actualContainerName)
+
+	if actualContainerName != containerName {
+		return fmt.Errorf("incorrect container name, expected %s, received %s",
+			containerName, actualContainerName)
+	}
+
+	if containerMetadataMap["DockerID"] == nil {
```
Comment (richardpen) [inline misc/agent-introspection-validator/agent-introspection-validator.go]: Shouldn't this be `DockerId`?
